### PR TITLE
[keyboard] Add Phantagom numpads

### PR DESCRIPTION
**Pull request by dkruyt**
## Description

Add support for 2 numpads / marco pads. Includes default and via support.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/phantagom/oled_mpad/config.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+#pragma once
+
+#include "config_common.h"
+
+#define VENDOR_ID 0xF8E8
+#define PRODUCT_ID 0x0001
+#define DEVICE_VER      0x0003
+#define MANUFACTURER    phantagom
+#define PRODUCT         mega numpad with oled and extra function rows
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/config.h:26]:
```suggestion
```

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    MO(_MS), KC_MUTE, KC_VOLD, KC_VOLU,
+    KC_PSCR, KC_NUM, KC_PSLS, KC_PAST, KC_PMNS, 
+    TO(_NP), KC_HOME,   KC_UP,   KC_PGUP,   
+    TO(_MS), KC_LEFT,   _______,   KC_RIGHT,   KC_PPLS, 
+    TO(_NV), KC_END,   KC_DOWN,   KC_PGDOWN,   
+    KC_A, KC_INS, KC_DEL,  KC_PENT  
+  ),
+};
+
+void matrix_init_kb(void) {
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:57]: User level functions should only be used in the keymap.
```suggestion
void matrix_init_user(void) {
```

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  writePinLow(GP4);
+  // set NumLock LED to output and low
+  setPinOutput(GP5);
+  writePinLow(GP5);
+  // set ScrollLock LED to output and low
+  setPinOutput(GP6);
+  writePinLow(GP6);
+}
+
+bool led_update_kb(led_t led_state) {
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:69]:
```suggestion
bool led_update_user(led_t led_state) {
```

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    }
+#endif 
+    return  true ;
+}
+
+void  oled_write_type_count ( void ) {
+     static  char type_count_str[ 7 ];
+     oled_write_P ( PSTR ( "Type count: " ), false );
+     itoa (type_count, type_count_str, 10 );
+     oled_write_ln (type_count_str, false );
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:99]:
```suggestion
     oled_write_ln(get_u16_str(type_count, ` `), false );
```

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    bool res = led_update_user(led_state);
+    if(res) {
+        writePin(GP4, led_state.num_lock);
+        writePin(GP5, led_state.caps_lock);
+        //writePin(B2, led_state.scroll_lock);
+    }
+    return res;
+}
+
+static  unsigned  int type_count = 0 ;
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:79]:
```suggestion
static uint16_t type_count = 0;
```

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+// write given digit, with leading zero if digit < 10
+void oled_write_2digit(unsigned int digit) {
+    static char buf[6];
+    itoa(digit, buf, 10);
+    if (digit < 10) {
+        oled_write_char('0', false);
+    }
+    oled_write(buf, false);
+}
+
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:144]: `get_u8_str` and `get_u16_str` will do this, and without using itoa, which can be expensive, size wise. 

if neither of these work, then you could simply use `get_numeric_str` instead. 

All of these are in quantum.c, and available anywhere.

File: keyboards/phantagom/oled_mpad/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,181 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/via/keymap.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/oled_mpad/keymaps/default/rules.mk [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# MCU name
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/rules.mk:1]: All of this should be removed, except what is needed (eg, what isn't present at the kb level).

File: keyboards/phantagom/oled_mpad/mcuconf.h [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#pragma once
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/mcuconf.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/oled_mpad/oled_mpad.c [hunk 11]
```diff
@@ -0,0 +1 @@
+#include "oled_mpad.h"
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/oled_mpad.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/oled_mpad/oled_mpad.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
+#pragma once
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/oled_mpad.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/rgb_npad/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,54 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/rgb_npad/keymaps/via/keymap.c [hunk 14]
```diff
@@ -0,0 +1,52 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/keymaps/via/keymap.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/rgb_npad/rgb_npad.c [hunk 15]
```diff
@@ -0,0 +1 @@
+#include "rgb_npad.h"
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/rgb_npad.c:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: keyboards/phantagom/rgb_npad/rgb_npad.h [hunk 16]
```diff
@@ -0,0 +1,39 @@
+#pragma once
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/rgb_npad.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 5 above)
Comment (dkruyt) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c]: needed some more work, but this issue is addressed

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 4 above)
Comment (dkruyt) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:85]: change to led_update_user compiles, but result in a not working keyboard. aka dead

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 7 above)
Comment (dkruyt) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:158]: I know what you mean, but not how to solve it, not a C programmer :( but learning a lot ;)

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 4 above)
Comment (joukewitteveen) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:85]: After #14452, this should probably become `void led_update_ports`.

Comment (drashna) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:85]: > change to led_update_user compiles, but result in a not working keyboard. aka dead

Did you remove the `bool res = led_update_user(led_state);` line below too?  If not, that probably caused a loop.

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 7 above)
Comment (drashna) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:158]: If you only wanted 2 digits: 
```c
const char *oled_write_2digit(uint8_t curr_num, char curr_pad) {
    static char    buf[3]   = {0};
    static uint8_t last_num = 0xFF;
    static char    last_pad = '\0';
    if (last_num == curr_num && last_pad == curr_pad) {
        return buf;
    }
    last_num = curr_num;
    last_pad = curr_pad;
    return get_numeric_str(buf, sizeof(buf), curr_num, curr_pad);
}
```

and `oled_write_2digits(uptime, '0')` would print out what you want.

File: keyboards/phantagom/oled_mpad/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+bool led_update_kb(led_t led_state) {
+    bool res = led_update_user(led_state);
+    if(res) {
+        writePin(GP4, led_state.num_lock);
+        writePin(GP5, led_state.caps_lock);
+        //writePin(B2, led_state.scroll_lock);
+    }
+    return res;
+}
```
Comment (joukewitteveen) [inline keyboards/phantagom/oled_mpad/keymaps/via/keymap.c:93]: Same as https://github.com/qmk/qmk_firmware/pull/18493/files#r990802288 :
```suggestion
void led_update_ports(led_t led_state) {
    writePin(GP4, led_state.num_lock);
    writePin(GP5, led_state.caps_lock);
    //writePin(B2, led_state.scroll_lock);
}
```
Note that this implementation nearly matches the default implementation. Why not set the right pin definitions (`LED_NUM_LOCK_PIN`, `LED_CAPS_LOCK_PIN`, but apparently not `LED_SCROLL_LOCK_PIN` or any others)?

File: keyboards/phantagom/oled_mpad/config.h [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED GP4
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_TIMEOUT 500U
+/* #define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED_MASK 0U */
+
+#define BOOTMAGIC_LITE_ROW 1
+#define BOOTMAGIC_LITE_COLUMN 0
+
+#define I2C_DRIVER I2CD1
+#define OLED_DISPLAY_128X64
+#define OLED_DISPLAY_ADDRESS 0x3c
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/config.h:55]: This looks to be the default address, so shouldn't need to be defined.
```suggestion
```

File: keyboards/phantagom/oled_mpad/keymaps/via/keymap.c (hunk 17 above)
Comment (dkruyt) [reply keyboards/phantagom/oled_mpad/keymaps/via/keymap.c:93]: changed to led_update_port and enabled scroll lock, dunno why I did not enabled it before...

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 7 above)
Comment (dkruyt) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:154]: This way I am getting extra leading 0, that I don't want, it's a uptime clock.

File: keyboards/phantagom/rgb_npad/readme.md [hunk 19]
```diff
@@ -0,0 +1,8 @@
+# RGB_NPAD
+
+![RGB_NPAD](https://kruyt.org/images/2022/09/rgb-numpad.jpg)
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/readme.md:3]: please rehost on imgur.

File: keyboards/phantagom/rgb_npad/info.json [hunk 20]
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "rgb_npad",
+    "keyboard_name": "rgb_npad",
+    "manufacturer": "Phantagom",
+    "vendorId": "0xF8E8",
+    "productId": "0x0002",
+    "matrix": { "rows": 6, "cols": 4 },
+    "url": "",
+    "lighting": "qmk_rgblight",
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/info.json:9]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/info.json [hunk 21]
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "rgb_npad",
+    "keyboard_name": "rgb_npad",
+    "manufacturer": "Phantagom",
+    "vendorId": "0xF8E8",
+    "productId": "0x0002",
+    "matrix": { "rows": 6, "cols": 4 },
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/info.json:7]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/info.json [hunk 22]
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "rgb_npad",
+    "keyboard_name": "rgb_npad",
+    "manufacturer": "Phantagom",
+    "vendorId": "0xF8E8",
+    "productId": "0x0002",
```
Comment (drashna) [inline keyboards/phantagom/rgb_npad/info.json:6]: These are not valid qmk configurations, but appear to be via json config. 
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c (hunk 7 above)
Comment (drashna) [reply keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:154]: `oled_write_2digits(uptime, ' ')` would fix that.

File: keyboards/phantagom/oled_mpad/keymaps/default/keymap.c [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+    static const char PROGMEM crkbd_logo[] = {
+        0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f, 0x90, 0x91, 0x92, 0x93, 0x94,
+        0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9, 0xaa, 0xab, 0xac, 0xad, 0xae, 0xaf, 0xb0, 0xb1, 0xb2, 0xb3, 0xb4,
+        0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7, 0xc8, 0xc9, 0xca, 0xcb, 0xcc, 0xcd, 0xce, 0xcf, 0xd0, 0xd1, 0xd2, 0xd3, 0xd4,
+        0};
+    oled_write_P(crkbd_logo, false);
+}
+
+#ifdef OLED_ENABLE
+bool oled_task_user(void) {
```
Comment (drashna) [inline keyboards/phantagom/oled_mpad/keymaps/default/keymap.c:180]: could all of this be moved to `oled_mpad.c`, and use the `oled_task_kb` function?

File: keyboards/phantagom/oled_mpad/keymaps/via/keymap.c (hunk 17 above)
Comment (waffle87) [inline keyboards/phantagom/oled_mpad/keymaps/via/keymap.c:93]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/config.h [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+#define TAPPING_TERM 400
+
+/* key matrix size */
+#define MATRIX_ROWS 6
+#define MATRIX_COLS 4
+
+// RGB
+#define RGB_DI_PIN GP28
+#define RGBLED_NUM 23
+#define NOP_FUDGE 0.4
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/config.h:31]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/config.h [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+#define LOCKING_SUPPORT_ENABLE
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_TIMEOUT 500U
+/* #define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED_MASK 0U */
+
+#define BOOTMAGIC_LITE_ROW 0
+#define BOOTMAGIC_LITE_COLUMN 0
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/config.h:74]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/info.json [hunk 26]
```diff
@@ -0,0 +1,44 @@
+{
+    "name": "rgb_npad",
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/info.json:2]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/readme.md [hunk 27]
```diff
@@ -0,0 +1,8 @@
+# RGB_NPAD
+
+![RGB_NPAD](https://i.imgur.com/EuGAMye.jpeg)
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/readme.md:3]:
```suggestion
![RGB_NPAD](https://i.imgur.com/EuGAMyeh.jpeg)
```

File: keyboards/phantagom/rgb_npad/rules.mk [hunk 28]
```diff
@@ -0,0 +1,20 @@
+# MCU name
+MCU = RP2040
+# Bootloader selection
+BOOTLOADER = rp2040
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/rules.mk:4]:
```suggestion
```

File: keyboards/phantagom/rgb_npad/info.json [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+    "keyboard_name": "rgb_npad",
+    "manufacturer": "Phantagom",
+    "url": "",
+    "maintainer": "Phantagom",
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/info.json:6]:
```suggestion
    "maintainer": "Phantagom",
    "processor": "RP2040",
    "bootloader": "rp2040",
```

File: keyboards/phantagom/rgb_npad/mcuconf.h [hunk 30]
```diff
@@ -0,0 +1,19 @@
+/*
```
Comment (waffle87) [inline keyboards/phantagom/rgb_npad/mcuconf.h:1]: This file should be removed.

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**